File: scala/src/virtual/abs/AbsGraph.scala
package virtual.abs

import scala.collection.mutable.HashSet

class AbsGraph(private var name: String) {

  type G <: AbsGraph
  type V <: AbsVertex
  type E <: AbsEdge

  def addVertex(v: V): Unit = {
    this.vertice.add(v)
    v.setGraph(this.asInstanceOf[v.G])
  }

  def removeVertex(v: V): Unit = {
    this.vertice.remove(v)
    v.unbindGraph()
  }

  def addEdge(e: E, v1: V, v2: V): Unit = {
    e.bindVertice(v1.asInstanceOf[e.V], v2.asInstanceOf[e.V])
    this.edges.add(e)
  }

  def removeEdge(e: E): Unit = {
    this.edges.remove(e)
    e.unbindGraph()
  }

  def getEdges: HashSet[E] = edges

  override def toString(): String = String.format("GŽnŽrique %s \n\tV = %s \n\tE = %s", name, vertice, edges).replace('[', '{').replace(']', '}')

  private var vertice: HashSet[V] = new HashSet[V]()
  private var edges: HashSet[E] = new HashSet[E]()
}

File: cpp/generics/abs/AbsGraph.cpp
/*
 * AbsGraph.cpp
 *
 *  Created on: 19 mai 2013
 *      Author: Abdelhamid
 */

#include "AbsGraph.h"

template <class G, class V, class E>
AbsGraph<G, V, E>::AbsGraph(string name)
{
	this->name = "Generic " + name;
	this->vertice = new set<V*>();
	this->edges = new set<E*>();
}


template <class G, class V, class E>
virtual void AbsGraph<G, V, E>::addVertex(V* v)
{
	this->vertice->insert(v);
	v->setGraph(this);
}

template <class G, class V, class E>
virtual void AbsGraph<G, V, E>::removeVertex(V* v)
{
	this->vertice->erase(v);
	v->unbindGraph();
}

template <class G, class V, class E>
virtual void AbsGraph<G, V, E>::addEdge(E* e, V* v1, V* v2)
{
	e->bindVertice(v1, v2);
	this->edges->insert(e);
}

template <class G, class V, class E>
virtual void AbsGraph<G, V, E>::removeEdge(E* e)
{
	this->edges->erase(e);
	e->unbindGraph();
}

template <class G, class V, class E>
virtual set<E*>* AbsGraph<G, V, E>::getEdges()
{
	return this->edges;
}
